Do not force internal collectors to call rrdset_next.

vkalintiris opened the pull request:
##### Summary

Removes a lot of boilerplate code. As explained in the commit messages, the only place that really needs the ability to call rrdset_next is the pluginsd parser. For every other in-process collector, timing should be an implementation detail.

This PR also removes `checks.plugins` while at it, mostly because I don't see how it'd offer any user-facing value and there's no need to maintain it.

##### Test Plan

- CI jobs

Dim-P commented on database/rrdset.c:
```diff
@@ -1342,12 +1342,15 @@ void rrdset_done(RRDSET *st) {
+void rrdset_timed_done(RRDSET *st, struct timeval now, bool pending_rrdset_next) {
     if(unlikely(netdata_exit)) return;
 
+    if (pending_rrdset_next)
```
This could do with a `likely()` wrapper.

vkalintiris replied:
Better to leave this as it is. The plugin's parser will set this to false.

Dim-P replied:
OK!

Dim-P commented on collectors/cgroups.plugin/sys_fs_cgroup.c:
```diff
@@ -2847,57 +2846,45 @@ void update_systemd_services_charts(
-        }
-        else
-            rrdset_next(st_cpu);
+    if (do_cpu && !st_cpu) {
```
Add `unlikely()` wrapper ?

Dim-P commented on collectors/cgroups.plugin/sys_fs_cgroup.c:
```diff
@@ -2847,57 +2846,45 @@ void update_systemd_services_charts(
-        }
-        else
-            rrdset_next(st_mem_usage);
+    if (do_mem_usage && !st_mem_usage) {
```
Same, add `unlikely()` wrapper ?

Dim-P commented on collectors/cgroups.plugin/sys_fs_cgroup.c:
```diff
@@ -3058,61 +3020,45 @@ void update_systemd_services_charts(
-        }
-        else
-            rrdset_next(st_mem_failcnt);
+    if(do_mem_failcnt && !st_mem_failcnt) {
```
Same, add `unlikely()` wrapper ?

Dim-P reviewed (commented):
@vkalintiris Nice PR, a couple of questions:

1. Most of the collectors would check in the first iteration if an `RRDSET *` exists and if not, they would call `rrdset_create_localhost()` instead of `rrdset_next()`. With this PR, it seems that `rrdset_next()` will also be called in the first iteration. I can't find anything wrong with that (I checked `rrdset_timed_next()` and it looks OK), just checking that it doesn't introduce any new issues indeed?
2. `rrdset_next()` was called before any `rrddim_set_by_pointer()` calls, but with this PR it is called after. Again, I can't notice any issues with that, just pointing out the obvious to confirm this change doesn't introduce any bugs?

I can't comment on `checks.plugin` as it's not something I've used, so I'll leave this to someone else (if somebody thinks it shouldn't be removed). 

Other than that, I am testing this and will approve if nothing seems off.

vkalintiris commented:
> @vkalintiris Nice PR, a couple of questions:
> 
> 1. Most of the collectors would check in the first iteration if an `RRDSET *` exists and if not, they would call `rrdset_create_localhost()` instead of `rrdset_next()`. With this PR, it seems that `rrdset_next()` will also be called in the first iteration. I can't find anything wrong with that (I checked `rrdset_timed_next()` and it looks OK), just checking that it doesn't introduce any new issues indeed?
> 2. `rrdset_next()` was called before any `rrddim_set_by_pointer()` calls, but with this PR it is called after. Again, I can't notice any issues with that, just pointing out the obvious to confirm this change doesn't introduce any bugs?
> 
> I can't comment on `checks.plugin` as it's not something I've used, so I'll leave this to someone else (if somebody thinks it shouldn't be removed).
> 
> Other than that, I am testing this and will approve if nothing seems off.

You're right, the reason I noticed that we can remove rrdset_next calls is that the same check can be done internally by testing `rrdset`'s `counter_done`. Rebased and fixed with the latest push.

Dim-P reviewed (dismissed):
Been testing this for a while, no stability or other issues noticed. LGMT!

thiagoftsm reviewed (changes requested):
Hello @vkalintiris ,

On FreeBSD we cannot compile this PR:

```sh
collectors/freebsd.plugin/freebsd_getifaddrs.c:590:7: error: expected expression
    } else {
      ^
collectors/freebsd.plugin/freebsd_getifaddrs.c:598:2: error: expected '}'
}
 ^
collectors/freebsd.plugin/freebsd_getifaddrs.c:143:48: note: to match this '{'
int do_getifaddrs(int update_every, usec_t dt) {
```

Please, fix this issue and also test on Mac.

vkalintiris commented:
> collectors/freebsd.plugin/freebsd_getifaddrs.c:598:2: error: expected '}'
> }
>  ^
> collectors/freebsd.plugin/freebsd_getifaddrs.c:143:48: note: to match this '{'
> int do_getifaddrs(int update_every, usec_t dt) {
> ```
> 
> Please, fix this issue and also test on Mac.

Fixed. Builds fine on MacOS.

thiagoftsm reviewed (approved):
I can confirm the PR is now compiling and running on `FreeBSD 13.1-RELEASE-p4` and on `Slackware current`. LGTM!

thiagoftsm reviewed (approved):
I retested on environments I already tested and everything worked, LGTM!
